File: result.go
package typesafe

// Usage is the token count for one call. A count is -1 when the API did
// not report it.
type Usage struct {
	InputTokens  int
	OutputTokens int
}

// Result is a fully decoded evaluation response.
type Result struct {
	// Model is the model that answered, for example "jev-1.13.0".
	Model string
	// Answers is keyed by the ids you chose. Use the typed accessors or a
	// type switch.
	Answers map[string]Answer
	Usage   Usage
	// RequestID is the x-typesafe-request-id header, for support requests.
	RequestID string
	// Raw is the decoded response body, including anything this package
	// does not model.
	Raw map[string]any
}

// Noul returns the Noul answer for id, or nil if absent or another type.
func (r *Result) Noul(id string) *NoulAnswer {
	a, _ := r.Answers[id].(*NoulAnswer)
	return a
}

// Choice returns the Choice answer for id, or nil if absent or another type.
func (r *Result) Choice(id string) *ChoiceAnswer {
	a, _ := r.Answers[id].(*ChoiceAnswer)
	return a
}

// Score returns the Score answer for id, or nil if absent or another type.
func (r *Result) Score(id string) *ScoreAnswer {
	a, _ := r.Answers[id].(*ScoreAnswer)
	return a
}

func decodeResult(body map[string]any, qs Questions) (*Result, error) {
	model, ok := body["model"].(string)
	if !ok {
		return nil, unexpectedError(body, `missing or invalid "model"`)
	}
	rawAnswers, ok := body["answers"].(map[string]any)
	if !ok {
		return nil, unexpectedError(body, `missing or invalid "answers"`)
	}
	rawUsage, ok := body["usage"].(map[string]any)
	if !ok {
		return nil, unexpectedError(body, `missing or invalid "usage"`)
	}
	answers := make(map[string]Answer, len(qs))
	for id, q := range qs {
		raw, ok := rawAnswers[id]
		if !ok {
			return nil, unexpectedError(rawAnswers, "missing answer for %q", id)
		}
		a, err := decodeAnswer(id, raw, q)
		if err != nil {
			return nil, err
		}
		answers[id] = a
	}
	return &Result{
		Model:   model,
		Answers: answers,
		Usage:   Usage{InputTokens: intField(rawUsage, "input_tokens"), OutputTokens: intField(rawUsage, "output_tokens")},
		Raw:     body,
	}, nil
}

File: lib/typesafe_api/result.ex
defmodule TypeSafeAPI.Result do
  @moduledoc """
  A fully decoded response: every answer keyed by the caller's own question
  ids, plus usage and the raw decoded body for anything this layer doesn't
  expose yet.

  Decoding is strict about the envelope (`model` must be present and a string,
  every question must have an answer, every answer must agree with its
  question) but permissive in two places: an answer the caller never asked for
  is dropped from `answers` and stays reachable through `raw`, and a missing or
  malformed `usage` decodes to a `TypeSafeAPI.Usage` with `nil` counts rather
  than failing a complete set of answers. The spec marks `usage` required with
  both token counts required, so the tolerance is defensive, not permission.

  `retry_count` is how many retries the call burned before this response (0 when
  the first attempt succeeded). `request_id` is the `x-typesafe-request-id` response header, the value
  TypeSafe support asks for. `TypeSafeAPI.Error` carries the same field for
  failed calls, including for a body that failed to decode here.

  `raw` holds the whole decoded body, so a result carries a second copy of
  everything `answers` already holds. `inspect/1` elides it: read
  `result.raw` directly when you want it.
  """

  alias TypeSafeAPI.{Answer, Error, Keys, Question, Usage}

  @type t :: %__MODULE__{
          model: String.t(),
          answers: %{Keys.key() => Answer.t()},
          usage: Usage.t(),
          request_id: String.t() | nil,
          retry_count: non_neg_integer(),
          raw: map()
        }

  @enforce_keys [:model, :answers, :usage, :raw]
  defstruct [:model, :answers, :usage, :raw, request_id: nil, retry_count: 0]

  @doc """
  Decodes a response body against the questions that were sent.

  `questions` is the normalized `[{id, question}]` list (see
  `TypeSafeAPI.Question.normalize/1`) and `keys` is the `TypeSafeAPI.Keys` built
  from it, so answers come back keyed and valued under the caller's own ids.

  `request_id` is the `x-typesafe-request-id` header of the response being
  decoded. It lands on the `TypeSafeAPI.Result` on success and on the
  `TypeSafeAPI.Error` on failure, so a decode mismatch — the one failure that
  needs a support ticket — can be quoted. `decode/3` passes `nil`.
  """
  @spec decode(map(), [{Keys.key(), Question.t()}], Keys.t(), String.t() | nil) ::
          {:ok, t()} | {:error, Error.t()}
  def decode(body, questions, keys, request_id \\ nil)

  def decode(body, questions, keys, request_id) when is_map(body) and is_list(questions) do
    with {:ok, model} <- fetch_model(body),
         {:ok, raw_answers} <- fetch_answers(body),
         {:ok, answers} <- decode_answers(raw_answers, questions, keys) do
      {:ok,
       %__MODULE__{
         model: model,
         answers: answers,
         usage: Usage.decode(Map.get(body, "usage")),
         request_id: request_id,
         raw: body
       }}
    else
      {:error, error} -> {:error, %{error | request_id: request_id}}
    end
  end

  defp fetch_model(%{"model" => model}) when is_binary(model), do: {:ok, model}
  defp fetch_model(body), do: {:error, Error.unexpected("missing or invalid \"model\"", body)}

  defp fetch_answers(%{"answers" => answers}) when is_map(answers), do: {:ok, answers}
  defp fetch_answers(body), do: {:error, Error.unexpected("missing or invalid \"answers\"", body)}

  defp decode_answers(raw_answers, questions, keys) do
    Enum.reduce_while(questions, {:ok, %{}}, fn {id, question}, {:ok, acc} ->
      wire_id = Keys.wire(id)

      with {:ok, raw_answer} <- fetch_raw_answer(raw_answers, wire_id, id),
           {:ok, answer} <- Answer.decode(raw_answer, wire_id, question, keys) do
        {:cont, {:ok, Map.put(acc, id, answer)}}
      else
        {:error, _reason} = error -> {:halt, error}
      end
    end)
  end

  defp fetch_raw_answer(raw_answers, wire_id, id) do
    case Map.fetch(raw_answers, wire_id) do
      {:ok, raw_answer} -> {:ok, raw_answer}
      :error -> {:error, Error.unexpected("missing answer for #{inspect(id)}", raw_answers)}
    end
  end

  defimpl Inspect do
    import Inspect.Algebra

    # `raw` is a second copy of everything above it, so printing a result would
    # otherwise print the whole response twice. Read `result.raw` for it.
    def inspect(result, opts) do
      fields = [
        model: to_doc(result.model, opts),
        answers: to_doc(result.answers, opts),
        usage: to_doc(result.usage, opts),
        request_id: to_doc(result.request_id, opts),
        raw: "#elided"
      ]

      container_doc("#TypeSafeAPI.Result<", fields, ">", opts, fn {key, doc}, _opts ->
        concat([Atom.to_string(key), ": ", doc])
      end)
    end
  end
end
